src/components/Router/HistoryRouter.tsx

```diff
@@ -0,0 +1,256 @@
+import { h, Component } from 'preact'
+import {
+  createMemoryHistory,
+  createBrowserHistory,
+  History,
+  LocationListener,
+  MemoryHistory,
+} from 'history'
+
+import { buildComponentsList } from './StepComponentMap'
+import StepsRouter from './StepsRouter'
+
+import { trackException } from '../../Tracker'
+
+import type { ApiParsedError, ErrorCallback } from '~types/api'
+import type { ExtendedStepTypes, FlowVariants } from '~types/commons'
+import type { CaptureKeys } from '~types/redux'
+import type {
+  ComponentStep,
+  ChangeFlowProp,
+  HistoryRouterProps,
+} from '~types/routers'
+import type { DocumentTypes, StepConfigDocument } from '~types/steps'
+
+type FormattedError = {
+  type: 'expired_token' | 'exception'
+  message: string
+}
+
+type HistoryLocationState = {
+  step: number
+  flow: FlowVariants
+}
+
+type State = {
+  initialStep: number
+} & HistoryLocationState
+
+const findFirstIndex = (
+  componentsList: ComponentStep[],
+  clientStepIndex: number
+) => componentsList.findIndex(({ stepIndex }) => stepIndex === clientStepIndex)
+
+export default class HistoryRouter extends Component<
+  HistoryRouterProps,
+  State
+> {
+  private history:
+    | MemoryHistory<HistoryLocationState>
+    | History<HistoryLocationState>
+  private unlisten: () => void
+
+  constructor(props: HistoryRouterProps) {
+    super(props)
+
+    const componentsList = this.getComponentsList(
+      { flow: 'captureSteps' },
+      this.props
+    )
+
+    const stepIndex =
+      this.props.stepIndexType === 'client'
+        ? findFirstIndex(componentsList, this.props.step || 0)
+        : this.props.step || 0
+
+    this.state = {
+      flow: 'captureSteps',
+      step: stepIndex,
+      initialStep: stepIndex,
+    }
+    this.history = this.props.options.useMemoryHistory
+      ? createMemoryHistory()
+      : createBrowserHistory()
+    this.unlisten = this.history.listen(this.onHistoryChange)
+    this.setStepIndex(this.state.step, this.state.flow)
+  }
+
+  onHistoryChange: LocationListener<HistoryLocationState> = ({
+    state: historyState,
+  }) => {
+    this.setState({ ...historyState })
+  }
+
+  componentWillUnmount(): void {
+    this.unlisten()
+  }
+
+  getStepType = (step: number): Optional<ExtendedStepTypes> => {
+    const componentList = this.getComponentsList()
+    return componentList[step] ? componentList[step].step.type : null
+  }
+
+  disableNavigation = (): boolean => {
+    return (
+      this.props.isNavigationDisabled ||
+      this.initialStep() ||
+      this.getStepType(this.state.step) === 'complete'
+    )
+  }
+
+  initialStep = (): boolean =>
+    this.state.initialStep === this.state.step &&
+    this.state.flow === 'captureSteps'
+
+  changeFlowTo: ChangeFlowProp = (
+    newFlow,
+    newStep = 0,
+    excludeStepFromHistory = false
+  ) => {
```

Comment:
Why we don't state what this function returns eg. `void`?

Reply:
Indeed the [return type](https://github.com/onfido/onfido-sdk-ui/blob/improvement/convert-routers-to-ts/src/types/routers.ts#L37-L41) for that function is `void`, but declaring `changeFlowTo: ChangeFlowProp = () => {}` gives types to all of its params, instead of delaring them again. This is called function signature.